### ISSUE TITLE
Unhandled exception when use Exiv2::ImageFactory::open()

### ISSUE THREAD

**Issue by GravitoWang** [Exiv2/exiv2#735]
Recent days,I was working on a project about extract GPS imformation from a photo. But I met a unexpected problem when I start it. The code is:
Unhandled exception at 0x00007FF9CD0EA388 in GUI_GPS.exe: Microsoft C++ exception: Exiv2::BasicError<char> at memory location 0x000000F808CF93E0.
And after some attempt to solve this problem, the exception code turned into:
0x00007FF98AD4B916 (exiv2.dll) (GUI_GPS.exe 中)Unhandled exception:  0xC0000005:  conflict occured when read 0xFFFFFFFFFFFFFFFF .
What's wrong with my code, please help me.I work on VisualStudio2013 with EXIV2 version 0.27.0, and my code is below:

``` C++
#include <iostream>
#include <exiv2\exiv2.hpp>
using namespace std;
using namespace Exiv2;

void main()
{
       //Exception here!
	//static unique_ptr<Exiv2::Image> image = ImageFactory::open("DJI_0010.JPG");
       //I change it into:
          decltype(auto) image = ImageFactory::open("DJI_0010.JPG");
      //still exception

	if (image.get() == 0)
	{
		cout << "read file error." << endl;
	}
	image->readMetadata();
	ExifData ed = image->exifData();

	if (ed.empty())
	{
		cout << "ed empty error!" << endl;
	}
	else
	{
		cout << ed["Exif.GPSInfo.GPSLatitude"] << endl;
	}
}
```

I don't know how to upload a picture, so if you need more imformation, you can contact me. I come from China, and I am a student who is studying for a master's degree.
I eagerly need you help, thank you very much!

**Comment by D4N** [comment]
I haven't tried your code yet, but there's a few things you can try:
first, remove the `static` from `image`, this is not a POD type so declaring it static does only add problems.
Then, exiv2 0.27 does not use C++11, so I'd recommend that you instead use the following:
``` C++
decltype(auto) image = ImageFactory::open("DJI_0010.JPG");
```

Provided that you are compiling with C++ >= 14, otherwise just use `auto` (here it doesn't matter, but `decltype(auto)` tends to be more robust than plain `auto`).

**Comment by GravitoWang** [comment]
@D4N Thank you very much for your advice. Just now, I modified my code follow your advice, and tried it on both VS2015(C++14) and VS2017(C++15), but the same exception occured. It seems that  ImageFactory::open() can not find the picture. Then I give it the full file directory and try again. The result is upset. Still, unhandled exception occured.
 Last year, I worked with EXIV2 version 0.26.0, the same code which performed well and gave the correct result. Due to the runtime library  type, I can not  append these code to my main project. So, I have to apply version 0.27.0.
I do not konw if my computer configuration is wrong, or it is a bug of VS.

**Comment by D4N** [comment]
@GravitoWang How did you obtain exiv2? Did you build it yourself, are you using the conan recipe or have you downloaded it from exiv2.org/GitHub?

As a first start: grab yourself a copy of Dr. Memory, run your target application via drmemory.exe and post the output (afaik it opens notepad with the report) here.

That might uncover something.

However, I suspect that there is some Windows specific mismatch between your compiled application and the exiv2 dll. I would recommend that you build exiv2 yourself and link your application against it via cmake.

I hope this helps a little. Unfortunately my Windows specific knowledge is rather limited, since I no longer have to use it at work.

**Comment by GravitoWang** [comment]
@D4N Follow your advice, I use drmemory to check my application, and the result is:

r. Memory version 1.11.0 build 2 built on Aug 29 2016 02:41:18
Dr. Memory results for pid 73900: "EXIV2015.exe"
Application cmdline: "D:\Project\EXIV2015\x64\Debug\EXIV2015.exe"
Recorded 115 suppression(s) from default E:\Program Files (x86)\Dr. Memory\bin64\suppress-default.txt

Error #1: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
replace_memchr 
    d:\drmemory_package\drmemory\replace.c(284):
exiv2.dll!std::_Traits_find<>
    ??:0
exiv2.dll!Exiv2::fileProtocol
    ??:0
exiv2.dll!Exiv2::ImageFactory::createIo
    ??:0
exiv2.dll!Exiv2::ImageFactory::open
    ??:0
ucrtbased.dll!set_app_type 
    ??:0
main        
    d:\project\exiv2015\exiv2015\exivtest.cpp(10):
Note: refers to 0 byte(s) beyond last valid byte in prior malloc

Error #2: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
replace_memchr 
    d:\drmemory_package\drmemory\replace.c(284):
exiv2.dll!std::_Traits_find<>
    ??:0
exiv2.dll!Exiv2::fileProtocol
    ??:0
exiv2.dll!Exiv2::ImageFactory::createIo
    ??:0
exiv2.dll!Exiv2::ImageFactory::open
    ??:0
ucrtbased.dll!set_app_type 
    ??:0
main        
    d:\project\exiv2015\exiv2015\exivtest.cpp(10):
Note: refers to 0 byte(s) beyond last valid byte in prior malloc

Error #3: LEAK 16 bytes 
replace_operator_new
    d:\drmemory_package\common\alloc_replace.c(2899):
std::_Allocate
    ??:0
KERNELBASE.dll!SetUnhandledExceptionFilter
    ??:0
ucrtbased.dll!set_app_type 
    ??:0
ucrtbased.dll!putwch_nolock
    ??:0
ucrtbased.dll!set_new_handler
    ??:0
ucrtbased.dll!initterm_e   
    ??:0
KERNEL32.dll!BaseThreadInitThunk
    ??:0

Error #4: LEAK 32 bytes 
replace_operator_new
    d:\drmemory_package\common\alloc_replace.c(2899):
std::_Allocate
    ??:0
std::allocator_traits<>::max_size
    ??:0
std::_Wrap_alloc<>::max_size
    ??:0
std::allocator<>::allocate
    ??:0
std::basic_string<>::max_size
    ??:0
std::_Wrap_alloc<>::allocate
    ??:0
std::_String_alloc<>::_Get_data
    ??:0
std::basic_string<>::_Copy
    ??:0
std::_String_alloc<>::_Mysize
    ??:0
std::basic_string<>::_Grow
    ??:0
std::_String_alloc<>::_Alloc_proxy
    ??:0

===========================================================================
FINAL SUMMARY:

DUPLICATE ERROR COUNTS:
	Error #   1:     10
	Error #   2:      5

SUPPRESSIONS USED:

ERRORS FOUND:
      2 unique,    15 total unaddressable access(es)
      0 unique,     0 total invalid heap argument(s)
      0 unique,     0 total GDI usage error(s)
      0 unique,     0 total handle leak(s)
      0 unique,     0 total warning(s)
      2 unique,     2 total,     48 byte(s) of leak(s)
      0 unique,     0 total,      0 byte(s) of possible leak(s)
Details: C:\Users\Gravitonium\AppData\Roaming\Dr. Memory\DrMemory-EXIV2015.exe.73900.000\results.txt

And, I find the function that errors occur :

void * __cdecl memchr (
        const void * buf,
        int chr,
        size_t cnt
        )
{
        //exception occurs here!
        while ( cnt && (*(unsigned char *)buf != (unsigned char)chr) ) {
                buf = (unsigned char *)buf + 1;
                cnt--;
        }

        return(cnt ? (void *)buf : NULL);
}

when exception thorw, some parameters are:

		buf	0x000001fd515de000	   const void *
		chr	104	                                   int
		cnt	18446697854098050410	   unsigned __int64

It seems that the application searches a character out of memory range, so exceptions are thowed out.
I do not konw what's wrong with it. This function is in file memchr.c, and is one of vcruntime files. Its full directory is E:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\crt\src\vcruntime\memchr.c

**Comment by D4N** [comment]
Hi @GravitoWang, sorry for the long delay. The error you are getting is rather odd, especially the zero bytes overreads.

How did you obtain exiv2? Did you build it yourself?

**Comment by GravitoWang** [comment]
@D4N Thank you  very much for your advice, I have solved the problem. It's about enviroment variable that  I have setted a excutable file directory(D:\exiv2-0.27.0-msvc64\bin) last year for EXIV2 which I directly downloaded form official website. And then I set another directory(D:\rebuild_exiv2_lib_2015\bin) this year  for EXIV2 which  I build it myself. Because, visual studio will always load .dll files firstly finded, so a wrongs exiv2.dll file is used. Thus, a seriers issues occur. Now, I have deleted the former directory and my code can correctly give a result.
At last, I apologize to you for my taking up your time with my own carelessness.
